src/micro.c

```diff
@@ -170,88 +246,145 @@ microGroup_AddEndpoint(microGroup *g, microEndpointConfig *cfg)
+    natsMutex_Lock(nc->servicesMu);
+    if (nc->services == NULL)
     {
-        _unlock_service(m);
-        return NULL;
+        nc->services = NATS_CALLOC(1, sizeof(microService *));
+        if (nc->services == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+    }
+    else
+    {
+        microService **tmp = NATS_REALLOC(nc->services, (nc->numServices + 1) * sizeof(microService *));
+        if (tmp == NULL)
+            s = nats_setDefaultError(NATS_NO_MEMORY);
+        else
+            nc->services = tmp;
     }
-    ep = m->first_ep;
 
-    for (; ep != NULL; ep = ep->next)
+    if (s == NATS_OK)
     {
-        if (err = micro_stop_endpoint(ep), err != NULL)
-        {
-            _unlock_service(m);
-            return microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->name);
-        }
+        nc->services[nc->numServices] = service;
+        nc->numServices++;
     }
+    natsMutex_Unlock(nc->servicesMu);
 
-    finalize = (m->first_ep == NULL);
-    if (finalize)
+    return micro_ErrorFromStatus(s);
+}
+
+static void
+_detach_service_from_connection(natsConnection *nc, microService *m)
+{
+    if (nc == NULL || m == NULL)
+        return;
+
+    natsMutex_Lock(nc->servicesMu);
+    for (int i = 0; i < nc->numServices; i++)
     {
-        natsLib_stopServiceCallbacks(m);
-        m->stopped = true;
-        doneHandler = m->cfg->DoneHandler;
+        if (nc->services[i] != m)
+            continue;
+        
+        for (int j = i; j < nc->numServices - 1; j++)
+            nc->services[j] = nc->services[j + 1];
+        nc->numServices--;
+        break;
     }
+    natsMutex_Unlock(nc->servicesMu);
 
-    _unlock_service(m);
+    return;
+}
+
+static microError *
+_stop_service(microService *m, bool unsubscribe, bool release)
+{
+    microError      *err            = NULL;
+    microEndpoint   *ep             = NULL;
+    int             refs            = 0;
+    int             numEndpoints    = 0;
 
-    if (finalize)
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    _lock_service(m);
+    if (!m->stopped)
     {
-        if (doneHandler != NULL)
-            doneHandler(m);
+        m->stopped = true;
 
-        // Relase the endpoint's server reference from `micro_add_endpoint`.
-        _release_service(m);
+        if (unsubscribe)
+        {
+            for (ep = m->first_ep; ep != NULL; ep = ep->next)
+            {
+                if (err = micro_stop_endpoint(ep), err != NULL)
+                {
+                    err = microError_Wrapf(err, "failed to stop service '%s', stopping endpoint '%s'", m->cfg->Name, ep->config->Name);
+                    _unlock_service(m);
+                    return err;
+                }
+            }
+        }
     }
 
+    if (release)
+        m->refs--;
+
+    refs = m->refs;
+    numEndpoints = m->numEndpoints;
+    _unlock_service(m);
+
+    if ((refs == 0) && (numEndpoints == 0))
```

Comment:
Question: would it be symptomatic of a bug if we have `refs == 0` and `numEndpoints != 0`?

Reply:
No, because of the order of execution. As I understand it (and can prove, I think) `On(Connection)Closed` may be invoked in the midst of Sub's threads calling the `On(Sub)Complete` from their own threads. `greyhair ++`